File: goldencheck/semantic/classifier.py
"""Semantic type classifier — infers what each column represents."""
from __future__ import annotations
import logging
from dataclasses import dataclass
from pathlib import Path
from typing import Any
import yaml
import polars as pl

logger = logging.getLogger(__name__)

@dataclass
class TypeDef:
    name_hints: list[str]
    value_signals: dict[str, Any]
    suppress: list[str]

@dataclass
class ColumnClassification:
    type_name: str | None
    source: str  # "name", "value", "llm", "none"

def load_type_defs(custom_path: Path | None = None) -> dict[str, TypeDef]:
    """Load base types + optional user types."""
    base_path = Path(__file__).parent / "types.yaml"
    with open(base_path) as f:
        base = yaml.safe_load(f)

    defs = {}
    for name, cfg in base.get("types", {}).items():
        defs[name] = TypeDef(
            name_hints=cfg.get("name_hints", []),
            value_signals=cfg.get("value_signals", {}),
            suppress=cfg.get("suppress", []),
        )

    # Merge user types (replace same-name, prepend new)
    if custom_path and custom_path.exists():
        with open(custom_path) as f:
            user = yaml.safe_load(f)
        for name, cfg in user.get("types", {}).items():
            defs[name] = TypeDef(
                name_hints=cfg.get("name_hints", []),
                value_signals=cfg.get("value_signals", {}),
                suppress=cfg.get("suppress", []),
            )

    return defs

def classify_columns(
    df: pl.DataFrame,
    custom_types_path: Path | None = None,
) -> dict[str, ColumnClassification]:
    """Classify each column's semantic type with provenance."""
    type_defs = load_type_defs(custom_types_path)
    results = {}

    for col_name in df.columns:
        # 1. Name heuristic matching
        classification = _match_by_name(col_name, type_defs)
        if classification:
            results[col_name] = ColumnClassification(type_name=classification, source="name")
            continue

        # 2. Value-based inference
        classification = _match_by_value(df, col_name, type_defs)
        if classification:
            results[col_name] = ColumnClassification(type_name=classification, source="value")
            continue

        results[col_name] = ColumnClassification(type_name=None, source="none")

    return results

def _match_by_name(col_name: str, type_defs: dict[str, TypeDef]) -> str | None:
    col_lower = col_name.lower()
    for type_name, type_def in type_defs.items():
        for hint in type_def.name_hints:
            # Prefix match: hint ends with _
            if hint.endswith("_"):
                if col_lower.startswith(hint):
                    return type_name
            # Suffix match: hint starts with _
            elif hint.startswith("_"):
                if col_lower.endswith(hint):
                    return type_name
            # Substring match (only for hints without _ prefix/suffix markers)
            elif hint in col_lower:
                return type_name
    return None

def _match_by_value(df: pl.DataFrame, col_name: str, type_defs: dict[str, TypeDef]) -> str | None:
    col = df[col_name]
    non_null = col.drop_nulls()
    if len(non_null) == 0:
        return None

    for type_name, type_def in type_defs.items():
        signals = type_def.value_signals
        if not signals:
            continue
        if _check_value_signals(non_null, col, signals):
            return type_name
    return None

def _check_value_signals(non_null: pl.Series, col: pl.Series, signals: dict) -> bool:
    """Check if ALL value signals are satisfied."""
    for key, value in signals.items():
        if key == "min_unique_pct":
            if non_null.n_unique() / len(non_null) < value:
                return False
        elif key == "max_unique":
            if non_null.n_unique() > value:
                return False
        elif key == "format_match":
            # Check format detection (email, phone, date)
            if not _check_format_match(non_null, value):
                return False
        elif key == "min_match_pct":
            pass  # Used with format_match, handled there
        elif key == "mixed_case":
            if non_null.dtype not in (pl.Utf8, pl.String):
                return False
            sample = non_null.head(100).to_list()
            has_upper = any(any(c.isupper() for c in str(s)) for s in sample)
            has_lower = any(any(c.islower() for c in str(s)) for s in sample)
            if not (has_upper and has_lower):
                return False
        elif key == "avg_length_min":
            if non_null.dtype not in (pl.Utf8, pl.String):
                return False
            avg_len = non_null.str.len_chars().mean()
            if avg_len is None or avg_len < value:
                return False
        elif key == "numeric":
            if col.dtype not in (pl.Int8, pl.Int16, pl.Int32, pl.Int64, pl.Float32, pl.Float64):
                return False
        elif key == "short_strings":
            if non_null.dtype not in (pl.Utf8, pl.String):
                return False
            avg_len = non_null.str.len_chars().mean()
            if avg_len is None or avg_len >= 5:
                return False
    return True

def _check_format_match(non_null: pl.Series, format_type: str) -> bool:
    if non_null.dtype not in (pl.Utf8, pl.String):
        return False
    if format_type == "email":
        matches = non_null.str.contains(r"@.*\.", literal=False).sum()
        return matches / len(non_null) >= 0.70
    elif format_type == "phone":
        matches = non_null.str.contains(r"\d{3}.*\d{3}.*\d{4}", literal=False).sum()
        return matches / len(non_null) >= 0.70
    elif format_type == "date":
        matches = non_null.str.contains(r"\d{4}-\d{2}-\d{2}", literal=False).sum()
        return matches / len(non_null) >= 0.50
    return False

File: packages/goldencheck-js/src/core/semantic/classifier.ts
/**
 * Semantic type classifier — classifies columns by name hints and value signals.
 * Port of goldencheck/semantic/classifier.py.
 */

import type { TabularData } from "../data.js";
import type { TypeDef, ColumnClassification } from "../types.js";
import { BASE_TYPES } from "./types.js";
import { getDomainTypes } from "./domains/index.js";

// Format-match regex patterns (matching Python's classifier)
const FORMAT_CHECKS: Record<string, { pattern: RegExp; minPct: number }> = {
  email: { pattern: /@.*\./, minPct: 0.70 },
  phone: { pattern: /\d{3}.*\d{3}.*\d{4}/, minPct: 0.70 },
  date: { pattern: /\d{4}-\d{2}-\d{2}/, minPct: 0.50 },
};

/**
 * Load type definitions with layered priority: base < domain < user.
 * Dict insertion order matters (later entries override earlier on name match).
 */
export function loadTypeDefs(
  domain?: string | null,
  userTypes?: Readonly<Record<string, TypeDef>>,
): Record<string, TypeDef> {
  const result: Record<string, TypeDef> = { ...BASE_TYPES };

  if (domain) {
    const domainTypes = getDomainTypes(domain);
    if (domainTypes) {
      Object.assign(result, domainTypes);
    }
  }

  if (userTypes) {
    Object.assign(result, userTypes);
  }

  return result;
}

/**
 * Classify all columns in a dataset.
 * Returns a map of column name → classification.
 */
export function classifyColumns(
  data: TabularData,
  typeDefs?: Record<string, TypeDef>,
  domain?: string | null,
): Record<string, ColumnClassification> {
  const defs = typeDefs ?? loadTypeDefs(domain);
  const result: Record<string, ColumnClassification> = {};

  for (const col of data.columns) {
    // Try name match first, then value match
    const nameMatch = matchByName(col, defs);
    if (nameMatch) {
      result[col] = nameMatch;
      continue;
    }

    const valueMatch = matchByValue(data, col, defs);
    if (valueMatch) {
      result[col] = valueMatch;
      continue;
    }

    result[col] = { typeName: null, source: "none" };
  }

  return result;
}

/**
 * Match column by name hints.
 * Hint ending with `_`: prefix-only match (NOT substring).
 * Hint starting with `_`: suffix-only match.
 * Otherwise: substring match.
 */
export function matchByName(
  colName: string,
  typeDefs: Readonly<Record<string, TypeDef>>,
): ColumnClassification | null {
  const lower = colName.toLowerCase();

  for (const [typeName, def] of Object.entries(typeDefs)) {
    for (const hint of def.nameHints) {
      const hintLower = hint.toLowerCase();

      if (hintLower.endsWith("_")) {
        // Prefix-only: "is_" matches "is_active" but NOT "diagnosis_desc"
        if (lower.startsWith(hintLower)) {
          return { typeName, source: "name" };
        }
      } else if (hintLower.startsWith("_")) {
        // Suffix-only: "_at" matches "created_at" but NOT "attention"
        if (lower.endsWith(hintLower)) {
          return { typeName, source: "name" };
        }
      } else {
        // Substring match
        if (lower.includes(hintLower)) {
          return { typeName, source: "name" };
        }
      }
    }
  }

  return null;
}

/**
 * Match column by value signals.
 * Returns first type where ALL signals pass.
 */
export function matchByValue(
  data: TabularData,
  colName: string,
  typeDefs: Readonly<Record<string, TypeDef>>,
): ColumnClassification | null {
  const nonNull = data.dropNulls(colName);
  if (nonNull.length === 0) return null;

  for (const [typeName, def] of Object.entries(typeDefs)) {
    if (Object.keys(def.valueSignals).length === 0) continue;
    if (checkValueSignals(data, colName, nonNull, def.valueSignals)) {
      return { typeName, source: "value" };
    }
  }

  return null;
}

function checkValueSignals(
  data: TabularData,
  colName: string,
  nonNull: unknown[],
  signals: Readonly<Record<string, unknown>>,
): boolean {
  const total = nonNull.length;

  for (const [key, value] of Object.entries(signals)) {
    switch (key) {
      case "min_unique_pct": {
        const uniquePct = data.nUnique(colName) / total;
        if (uniquePct < (value as number)) return false;
        break;
      }
      case "max_unique": {
        if (data.nUnique(colName) > (value as number)) return false;
        break;
      }
      case "format_match": {
        const fmtName = value as string;
        const fmt = FORMAT_CHECKS[fmtName];
        if (!fmt) break;
        const minPct = (signals["min_match_pct"] as number | undefined) ?? fmt.minPct;
        const strings = data.stringValues(colName);
        if (strings.length === 0) return false;
        const matchCount = strings.filter((s) => fmt.pattern.test(s)).length;
        if (matchCount / strings.length < minPct) return false;
        break;
      }
      case "min_match_pct":
        // Handled by format_match
        break;
      case "mixed_case": {
        const sample = data.stringValues(colName).slice(0, 50);
        const hasUpper = sample.some((s) => /[A-Z]/.test(s));
        const hasLower = sample.some((s) => /[a-z]/.test(s));
        if (!(hasUpper && hasLower)) return false;
        break;
      }
      case "avg_length_min": {
        const strings = data.stringValues(colName);
        if (strings.length === 0) return false;
        const avgLen = strings.reduce((sum, s) => sum + s.length, 0) / strings.length;
        if (avgLen < (value as number)) return false;
        break;
      }
      case "numeric": {
        if (!data.isNumeric(colName)) return false;
        break;
      }
      case "short_strings": {
        const strings = data.stringValues(colName);
        if (strings.length === 0) return false;
        const avgLen = strings.reduce((sum, s) => sum + s.length, 0) / strings.length;
        if (avgLen >= 5) return false;
        break;
      }
    }
  }

  return true;
}
